doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,435 @@
+from the stream send buffers.  The stream send buffers will return a
+constructed frame header and a pointer to the steam data and length.  A second
+call exists to allow the packetiser to know how much data is queued for a stream
+so that planning for the creation of multiple packets is possible.
+
+```c
+int ossl_quic_get_app_data(QUIC_STREAM *stream, size_t request,
+                           const OSSL_QUIC_FRAME **frame,
+                           const unsigned char **data,
+                           size_t *data_len);
+
+size_t ossl_quic_get_app_data_size(QUIC_STREAM *stream);
+```
+
+#### Notes
+
+* Unclear how to best free the data after sent data was acked.
+  The data will be fragments from the buffers so the stream send buffers will
+  need to remember which fragment have been sent and which are pending and
+  only free once everything is sent:
+
+```c
+int ossl_quic_free_app_data(QUIC_STREAM *stream, void *data, size_t data_len);
+```
+
+* Also need a call to tell the stream send buffers to forget about previously
+  requested app data because it needs to be retransmitted and the boundaries
+  could change.  This is similar to the free call but the data is placed/queue
+  back in the send buffers appropriately, hence it might need the stream offset as
+  well:
+
+```c
+int ossl_quic_retransmitting_app_data(QUIC_STREAM *stream, size_t offset,
+                                      void *data, size_t data_len);
+```
+
+### TLS Handshake Record Layer
+
+Uses the Record Layer API to implement the inner TLS-1.3 protocol handshake.
+It produces the QUIC crypto frames which are queued using the same mechanism
+as the [Stream Send Buffers](#stream-send-buffers) above.
+
+### Flow Controller and Statistics Collector
+
+To make decisions about what frames to coalesce, the packetiser relies
+on the flow controller to enforce stream and connection bandwidth limits
+[RFC 9000 4.1 Data Flow Control].
+
+```c
+/*
+ * Return the maximum amount of data that is permitted for the given stream.
+ * This includes both the stream limit and it's associated connection limit.
+ */
+size_t ossl_quic_stream_flow_maximum_size(QUIC_STREAM *stream);
+
+/*
+ * Information the flow controller that an amount of data has been queued for
+ * sending to a stream.
+ */
+int ossl_quic_stream_flow_sent_data(QUIC_STREAM *stream, size_t bytes);
+```
+
+### Congestion Controller
+
+Also part of the frame coalescing decision is the congestion controller
+[RFC 9002].  For MVP, this will be a _just send it_.
+
+```c
+/*
+ * Pluggable congestion controller APIs go here
+ * Extract that is required from #18018
+ */
+```
+
+### QUIC Write Record Layer
+
+Coalesced frames are passed to the QUIC record layer for encryption and sending.
+To send accumulated frames as packets to the QUIC Write Record Layer:
+
+```c
+int ossl_qtx_write_pkt(OSSL_QTX *qtx, const OSSL_QTX_PKT *pkt);
+```
+
+The packetiser will attempt to maximise the number of frames in a packet.
```

Comment:
I'm not sure maximising the *number* of frames in the packet is the goal. Isn't it more about ensuring the packet is as full as possible? This could be done with a small number of large frames if appropriate.

Reply:
And then we're into the bin packing problem which I'm not even going to suggest we try to solve.
There will be a lot more subtle points -- avoid no-ack and non-flow controlled frames and send them separately e.g.

I've changed it to _number of bytes_.